src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,204 @@
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private T unlink(Node<T> nodeToErase) {
+        final T removed = nodeToErase.value;
+        final Node<T> next = nodeToErase.next;
+        final Node<T> prev = nodeToErase.prev;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            nodeToErase.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            nodeToErase.next = null;
+        }
+        nodeToErase.value = null;
+        size--;
+        return removed;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (first == null);
+    }
+
+    private Node<T> node(int index) {
+        checkElementIndex(index);
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+
+            }
+            return x;
+        }
+    }
+
+    private void linkFirst(T value) {
+        Node<T> nextAfter = first;
+        Node<T> newFirst = new Node<>(null, value, nextAfter);
+        first = newFirst;
+        if (nextAfter == null) {
+            last = newFirst;
+        } else {
+            nextAfter.prev = newFirst;
+        }
+        size++;
+    }
+
+    private void linkLast(T value) {
+        Node<T> oldLast = last;
+        Node<T> addedNode = new Node(oldLast, value, null);
+        last = addedNode;
+        if (oldLast == null) {
+            first = addedNode;
+        } else {
+            oldLast.next = addedNode;
+        }
+        size++;
+    }
+
+    private boolean elementExists(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private boolean positionAddable(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!positionAddable(index)) {
+            throw new IndexOutOfBoundsException("index " + index + " size " + size);
+        }
+
+    }
```

Comment:
I dont think we need separate positionAddable method, lets remove this method and impltmtnr it's logic in checkPositionIndex method (same for checkElementIndex and elementExists)